test/unit/specs/components/v-button.spec.js

```diff
@@ -0,0 +1,125 @@
+import Vue from 'vue'
+import { render, screen } from '@testing-library/vue'
+
+import VButton from '~/components/VButton.vue'
+
+/**
+ * Throughout this suite we use the `screen.findBy*` functions to asynchronously
+ * wait for the component to re-render. There might be some kind of performance
+ * problem with the component's implementation, but if we don't "wait" for it
+ * to settle, then all the props that are changed after `onMounted` completes
+ * won't be rendered.
+ */
+describe('VButton', () => {
+  it('should render a `button` by default with type="button" and no tabindex', async () => {
+    render(VButton, {
+      slots: { default: 'Code is Poetry' },
+    })
+
+    const element = await screen.findByRole('button')
+
+    expect(element.tagName).toBe('BUTTON')
+    expect(element).toHaveAttribute('type', 'button')
+    expect(element).not.toHaveAttribute('tabindex')
+  })
+
+  it('should allow passing an explicit type', async () => {
+    render(VButton, {
+      props: { type: 'submit' },
+      slots: { default: 'Code is Poetry' },
+    })
+
+    const element = await screen.findByRole('button')
+
+    expect(element).toHaveAttribute('type', 'submit')
+  })
+
+  it('should allow passing in an explicit tabindex', async () => {
+    render(VButton, {
+      props: { tabindex: '-1' },
+      slots: { default: 'Code is Poetry' },
+    })
+
+    const element = await screen.findByRole('button')
+
+    expect(element).toHaveAttribute('tabindex', '-1')
+  })
+
+  it.each(['button', 'submit', 'reset', 'color', 'file', 'image'])(
+    'should allow an input element that renders a button with the type %s',
+    async (type) => {
+      const TestWrapper = Vue.component('TestWrapper', {
+        components: { VButton },
+        template: `<label>Code is Poetry<VButton as="input" type="${type}" /></label>`,
+      })
+      render(TestWrapper)
+
+      const element = await screen.findByLabelText(/code is poetry/i)
+
+      expect(element.tagName).toBe('INPUT')
+      expect(element).toHaveAttribute('type', type)
+    }
+  )
+
+  // @todo(sarayourfriend) fix this failing test!
+  it.skip('should render an anchor with no type attribute', async () => {
+    render(VButton, {
+      props: { as: 'a' },
+      slots: { default: 'Code is Poetry' },
+    })
+
+    const element = await screen.findByText(/code is poetry/i)
+
+    screen.debug(element)
+
+    expect(element.tagName).toBe('A')
+    expect(element).not.toHaveAttribute('type')
+  })
+
+  it('should render as a div with role and tabindex', async () => {
```

Comment:
Do we really want to allow creating a 'div' button? What would be the use case?

Reply:
Hmm, I'm not sure. It was necessary to cover because of how lax the `as` type is, it's just a pass through to the `is` prop on `Component` and so could be anything.

We could narrow it though and require the `as` prop to only be either `button` or `a` instead of handling every possible thing. That would limit certain scenarios like if you needed to pass another Vue component to `as`. In those cases we can't even properly do the validation because the `ref` wouldn't resolve to the HTML node anyway (I'm just realizing this is a problem in implementing this for Vue as there is no ref forwarding in Vue).

Probably we should just restrict this to either a basic anchor or button element instead of allowing any Vue component or HTML element...

I'll push an update for that.